lib/elixir/lib/calendar/duration.ex

```diff
@@ -197,4 +197,144 @@ defmodule Duration do
       microsecond: {-ms, p}
     }
   end
+
+  @doc """
+  Parses an ISO 8601-2 formatted duration string to a `Duration` struct.
+
+  ## Examples
+
+      iex> Duration.from_iso8601("P1Y2M3DT4H5M6S")
+      {:ok, %Duration{year: 1, month: 2, day: 3, hour: 4, minute: 5, second: 6}}
+      iex> Duration.from_iso8601("PT10H30M")
+      {:ok, %Duration{hour: 10, minute: 30, second: 0}}
+      iex> Duration.from_iso8601("P3Y-2MT3H")
+      {:ok, %Duration{year: 3, month: -2, hour: 3}}
+      iex> Duration.from_iso8601("-P3Y2MT3H")
+      {:ok, %Duration{year: -3, month: -2, hour: -3}}
+      iex> Duration.from_iso8601("-P3Y-2MT3H")
+      {:ok, %Duration{year: -3, month: 2, hour: -3}}
+
+  """
+  @spec from_iso8601(String.t()) :: {:ok, t} | {:error, String.t()}
+  def from_iso8601("P" <> duration_string) do
```

Comment:
Looks like ISO8601 duration has a `P<date>T<time> ` notation (Wikipedia's article give the following example: `P0003-06-04T12:30:05`). Do we intend to support it?

We could easily match on it as it is done in `Calendar.ISO`, especially if we move this code to this module as suggested by Jose. Something like:

```elixir
  # in calendar.ISO
  defp do_parse_duration(
         <<"P", unquote(match_basic_date), rest::binary>> = duration_with_datetime,
         other,
         params
       )
       when unquote(guard_date) do
       "P" <> datetime = duration_with_datetime
       with {:ok, {year, month, day, hour, minute, second, microsecond}} = parse_native_datetime(datetime) do
          # convert to duration
       end
  end
```

Reply:
I would avoid it for now and start with a small and consistent subset. :)